File: backend_flask/services/judge0_service.py
import base64
import json
import os
import time
from urllib.error import HTTPError
from urllib.request import Request, urlopen

DEFAULT_JUDGE0_BASE_URL = "https://ce.judge0.com"
LANGUAGE_IDS = {
    "c": 50,
    "cpp": 54,
    "java": 62,
    "python": 71,
    "rust": 73,
}
TEXT_RESULT_FIELDS = ["stdout", "stderr", "compile_output", "message"]


def _encode_base64(value):
    return base64.b64encode(str(value or "").encode("utf-8")).decode("ascii")


def _decode_base64(value):
    if not isinstance(value, str) or not value:
        return value

    return base64.b64decode(value.encode("ascii")).decode("utf-8")


def _decode_judge0_text_fields(result):
    if not isinstance(result, dict):
        return result

    decoded = {**result}

    for field in TEXT_RESULT_FIELDS:
        decoded[field] = _decode_base64(decoded.get(field))

    return decoded


def _get_judge0_base_url():
    return (os.getenv("JUDGE0_BASE_URL") or DEFAULT_JUDGE0_BASE_URL).rstrip("/")


def _get_judge0_headers():
    headers = {
        "Content-Type": "application/json",
    }

    auth_token = os.getenv("JUDGE0_AUTH_TOKEN")
    rapidapi_key = os.getenv("JUDGE0_RAPIDAPI_KEY")

    if auth_token:
        headers["X-Auth-Token"] = auth_token

    if rapidapi_key:
        headers["X-RapidAPI-Key"] = rapidapi_key
        headers["X-RapidAPI-Host"] = (
            os.getenv("JUDGE0_RAPIDAPI_HOST") or "judge0-ce.p.rapidapi.com"
        )

    return headers


def _parse_judge0_response(response):
    raw_data = response.read().decode("utf-8")

    try:
        return json.loads(raw_data or "{}")
    except json.JSONDecodeError:
        return {}


def _request_json(url, headers, method="GET", payload=None):
    body = None

    if payload is not None:
        body = json.dumps(payload).encode("utf-8")

    request = Request(url, data=body, headers=headers, method=method)

    try:
        with urlopen(request, timeout=30) as response:
            return _parse_judge0_response(response)
    except HTTPError as error:
        data = _parse_judge0_response(error)
        message = (
            data.get("error")
            or data.get("message")
            or f"Judge0 request failed with status {error.code}."
        )
        raise RuntimeError(message) from error


def get_language_id(language):
    return LANGUAGE_IDS.get(language)


def run_judge0_submission(
    language,
    source_code,
    stdin="",
    time_limit_seconds=1,
    memory_limit_mb=256,
):
    language_id = get_language_id(language)

    if not language_id:
        raise RuntimeError("Unsupported language.")

    base_url = _get_judge0_base_url()
    headers = _get_judge0_headers()
    created = _request_json(
        f"{base_url}/submissions?base64_encoded=true&wait=false",
        headers,
        method="POST",
        payload={
            "language_id": language_id,
            "source_code": _encode_base64(source_code),
            "stdin": _encode_base64(stdin),
            "cpu_time_limit": float(time_limit_seconds or 1),
            "memory_limit": int(memory_limit_mb or 256) * 1024,
        },
    )

    token = created.get("token")

    if not token:
        return _decode_judge0_text_fields(created)

    for attempt in range(25):
        time.sleep(0.5 if attempt < 3 else 1)
        result = _request_json(
            f"{base_url}/submissions/{token}?base64_encoded=true",
            headers,
        )
        status = result.get("status") or {}

        try:
            status_id = int(status.get("id") or 0)
        except (TypeError, ValueError):
            status_id = 0

        if status_id > 2:
            return _decode_judge0_text_fields(result)

    raise RuntimeError("Judge0 submission timed out while waiting for a result.")

File: backend/src/services/judge0.service.js
const DEFAULT_JUDGE0_BASE_URL = "https://ce.judge0.com";
const LANGUAGE_IDS = {
  c: 50,
  cpp: 54,
  java: 62,
  python: 71,
  rust: 73,
};
const TEXT_RESULT_FIELDS = ["stdout", "stderr", "compile_output", "message"];

function encodeBase64(value) {
  return Buffer.from(String(value || ""), "utf8").toString("base64");
}

function decodeBase64(value) {
  if (typeof value !== "string" || value.length === 0) {
    return value;
  }

  return Buffer.from(value, "base64").toString("utf8");
}

// takes a Judge0 result object and decodes its base64 text fields like stdout, stderr, and compile_output
function decodeJudge0TextFields(result) {
  if (!result || typeof result !== "object") {
    return result;
  }

  return TEXT_RESULT_FIELDS.reduce(
    (acc, field) => ({
      ...acc,
      [field]: decodeBase64(acc[field]),
    }),
    { ...result },
  );
}

function getJudge0BaseUrl() {
  return (process.env.JUDGE0_BASE_URL || DEFAULT_JUDGE0_BASE_URL).replace(
    /\/+$/,
    "",
  );
}

function getJudge0Headers() {
  const headers = {
    "Content-Type": "application/json",
  };

  if (process.env.JUDGE0_AUTH_TOKEN) {
    headers["X-Auth-Token"] = process.env.JUDGE0_AUTH_TOKEN;
  }

  if (process.env.JUDGE0_RAPIDAPI_KEY) {
    headers["X-RapidAPI-Key"] = process.env.JUDGE0_RAPIDAPI_KEY;
    headers["X-RapidAPI-Host"] =
      process.env.JUDGE0_RAPIDAPI_HOST || "judge0-ce.p.rapidapi.com";
  }

  return headers;
}

function wait(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

async function parseJudge0Response(response) {
  const data = await response.json().catch(() => ({}));

  if (!response.ok) {
    const message =
      data?.error ||
      data?.message ||
      `Judge0 request failed with status ${response.status}.`;
    throw new Error(message);
  }

  return data;
}

export function getLanguageId(language) {
  return LANGUAGE_IDS[language] || null;
}

// sends code to Judge0, waits until Judge0 finishes judging it, then returns the decoded result
export async function runJudge0Submission({
  language,
  sourceCode,
  stdin = "",
  timeLimitSeconds = 1,
  memoryLimitMb = 256,
}) {
  const languageId = getLanguageId(language);

  if (!languageId) {
    throw new Error("Unsupported language.");
  }

  const baseUrl = getJudge0BaseUrl();
  const headers = getJudge0Headers();
  const createResponse = await fetch(
    `${baseUrl}/submissions?base64_encoded=true&wait=false`,
    {
      method: "POST",
      headers,
      body: JSON.stringify({
        language_id: languageId,
        source_code: encodeBase64(sourceCode),
        stdin: encodeBase64(stdin),
        cpu_time_limit: Number(timeLimitSeconds) || 1,
        memory_limit: (Number(memoryLimitMb) || 256) * 1024,
      }),
    },
  );
  const created = await parseJudge0Response(createResponse); 
  if (!created.token) {
    return decodeJudge0TextFields(created);
  }

  for (let attempt = 0; attempt < 25; attempt += 1) {
    await wait(attempt < 3 ? 500 : 1000);

    const resultResponse = await fetch(
      `${baseUrl}/submissions/${created.token}?base64_encoded=true`,
      { headers },
    );
    const result = await parseJudge0Response(resultResponse);
    const statusId = Number(result.status?.id);

    if (statusId && statusId > 2) {
      return decodeJudge0TextFields(result);
    }
  }

  throw new Error("Judge0 submission timed out while waiting for a result.");
}
